### ISSUE TITLE
Issue with hooks

### ISSUE THREAD

**Issue by mattapperson** [keystonejs/grappling-hook#15]
OK so in this example:

``` js
instance.pre('initPlugins', function (foo, next) {
        foo.push('funky');
        next(null, foo)
})
instance.pre('initPlugins', function (foo, next) {
        foo.push('monkey');
        next(null, foo)
});

instance.callHook('pre:initPlugins', [
     'basic-plugin-1'
], function () {
    // arguments = { '0': undefined }
});
```

Not the result I expected. This next example is close, but still not the expected behavior:

``` js
instance.pre('initPlugins', function (foo, next) {
        foo.push('funky');
        next(null, foo)
})
instance.pre('initPlugins', function (foo, next) {
        foo.push('monkey');
        next(foo)
});

instance.callHook('pre:initPlugins', [
     'basic-plugin-1'
], function () {
    // arguments = { '0': [ 'basic-plugin-1', 'funky', 'monkey' ] }
});
```

My issue is, it seems hooks can pass data between each other, but not back to the calling code. Only errors seem to make it back. Is this intended? A Bug? Or am I missing something in the docs?

**Comment by mattapperson** [comment]
I might also add that this (what I first tried) short circuits the entire thing:

``` js
instance.pre('initPlugins', function (foo, next) {
        foo.push('funky');
        next(foo)
})
instance.pre('initPlugins', function (foo, next) {
        foo.push('monkey');
        next(foo)
});

instance.callHook('pre:initPlugins', [
     'basic-plugin-1'
], function () {
    // arguments = { '0': [ 'basic-plugin-1', 'funky' ] }
});
```

**Comment by mattapperson** [comment]
OK seems perhaps this is a deficiency with callHook. wrapped methods allow passing values to the final callback, but callHook does not...

**Comment by mattapperson** [comment]
At the end of the day what I would expect here is:

``` js
instance.pre('initPlugins', function (foo, next) {
        foo.push('funky');
        next(foo) //pass params through to the next hook
})
instance.pre('initPlugins', function (foo, next) {
        // This hook should still be called
        foo.push('monkey');
        next(foo)
});

instance.callHook('pre:initPlugins', [
     'basic-plugin-1'
], function () {
    // arguments = { '0': undefined', '1': [ 'basic-plugin-1', 'funky', 'monkey' ] }
});
```

where argument 0 is an error arg as a catch for any errors thrown in hooks, and everything after that is the arguments augmented by the hooks.

**Comment by mattapperson** [comment]
I would also add, I would expect a post hook to know what the results of the called method are...

**Comment by mattapperson** [comment]
Is there a use-case for post events needing a callback? 
In my mind that method sig might be better as (err, <results params>)

**Comment by creynders** [comment]
It's not a bug, it's a feature. :grin:
Seriously though, there's a really good reason why it behaves as it does. I'll try to explain.

First of all though: there's a difference between a node-style callback (with signature `function(err, a, b, c)`) and a middleware callback. The latter - as popularised by express - assumes any parameter passed to it is an error.
E.g. a typical express middleware function:

``` js
var requestTime = function (req, res, next) {
  req.requestTime = Date.now();
  next();
};
```

I.e. middleware manipulate any of their parameters, but don't pass anything on. (Except errors, should they occur)

So, let's get to the reasoning of why nothing is passed to the final callback in your above examples.

One of the major use cases of grappling-hook is to allow to enhance an object's method with hookability. I always aim to write libs that add functionality w/o the need to rewrite anything.
So, let's say there's a module `foo`, with a `save` method. Since `foo` is incredible useful, you use it in your application literally everywhere, like this:

``` js
//ex. A
var foo = require('foo');

foo.save(someJSON, function(success){
    if(success){
        console.log('JSON saved successfully!');
    }else{
        console.log('JSON could not be saved!!!');
    }
});
```

However requirements change and at a certain point in time you want to "hook" into the save process, e.g. you want to log the JSON being saved. One obvious solution would be to track down all of the `foo.save` call points and add the code which logs the json. 

Or you could use grappling-hook.

``` js
//ex. B
var foo = require('foo');
var grappling = require('grappling-hook');

grappling.mixin(foo);
grappling.addHooks('save');

foo.pre('save', function(json, next){
    //log json
    next();
});
```

That's it. Nothing else needs to change, and the code in ex. A does not need to change, even better; whatever the hell any middleware (added to the `pre:save` or `post:save` hooks) does, it won't interfere with any ex. A-like code. Since if it would, all hell breaks loose. It would break not only your application code, but potentially that of any dependents that require `foo` too.

However, you do have a point. I have to admit that with `callHook` it _is_ unintuitive. I always aim for symmetry, but maybe - maybe! - this is a valid case where symmetry could be broken.

Then, answers to some of your specific points:

> I would expect a post hook to know what the results of the called method are

Possibly. That's discussed in #12

> Is there a use-case for post events needing a callback? 

Yes, of course. To allow several asynchronous post middleware functions that are called in series, not parallel.

> In my mind that method sig might be better as (err, )

No. Any middleware handling is immediately aborted when an error is passed to a callback, i.e. `post` middleware will never be run if a `pre` middleware passes an error. However, maybe it's an idea to allow explicitly error handling middleware as well, just as in express.

**Comment by mattapperson** [comment]
@creynders thanks for the lengthy feedback, it's super helpful to understand your headspace 😊 
I think part of my issue is I am coming from the Hapi/plugin mindset and you are coming from the express/middleware mindset.

Anyway... So for your argument that it's not an argument it's a feature... My issue is that why then does it work for calling a wrapped function just not callHook.

As for the postHook here is what I see as a very valid use case:
Let's say I have a method in my code that pulls all my "friends" from the database in a social app. I want to hook that so that and add to the results all my Facebook friends and return them in 1 array. With the current API there is no way I can see to do this. If the post hook knew the results (and counter to my orig point that there is no use for asynchronous post hooks) could inject the Facebook friends into the results.

Most importantly then all the above is: are you averse flat out to the idea that a plugin can affect the greater system at large? If so, then I'll stop making arguments for things that it should  😊

**Comment by creynders** [comment]
> My issue is that why then does it work for calling a wrapped function just not callHook.

But it doesn't. Not like you think it does. The callback of the wrapped method just receives the same parameters as if the method would be non-wrapped. But with a `callHook` there's nothing to pass on, since there's no wrapped method. Obviously I could inject the original parameters passed to the middleware into the final callback of the `callHook` but that's rather useless IMO, since it can access them from the closure anyway. I.e. I don't see any sense in this:

``` js
instance.callHook('pre:whatever', foo, bar, baz, qux, function(foo, bar, baz, qux){
  //silly, no?
});
```

Anyway, AFAICT there's no use case you can't solve, you might just need to go about it a bit differently from what you expect. Except for the following of course:

> As for the postHook here is what I see as a very valid use case

Yeah, that's one of the use cases I had in mind too. Any further discussion on this is best placed in #12 

> are you averse flat out to the idea that a plugin can affect the greater system at large

Definitely. However, I don't tend to think about this as plugins. And it depends on what scope of "affect" you have in mind. Should a plugin/middleware be encapsulated as much as possible, to reduce the amount of possible side effects? Definitely.
Should it be allowed to modify outgoing data? Definitely, otherwise it'd be nigh useless.

**Comment by mattapperson** [comment]
@creynders For your example for callHook, I agree it might look a little silly at first glance, but the issue is that without that, within the closure you have no idea what the params might have been changed to by hooks the way things currently work. Wrapped methods do get their params altered by hooks. 
Simply put, in `callHook` there is no way to get access to data from hooks... that feels like a bug to me.

As for not affecting the system at large. I would only submit this as a thought: Wordpress lets you do this, and it has contributed in a major way to the power of using it as a platform. There are a million reasons why it's a bad idea, I agree. But at the same time, having that power available is, well, powerful.

**Comment by creynders** [comment]
> Simply put, in callHook there is no way to get access to data from hooks... that feels like a bug to me.

That's a misunderstanding. Let me clarify:

``` js
instance.pre('whatever', function(foo, next){
  foo.name="changed by middleware";
  next();
});

var foo = {
  name: 'foo'
};
console.log(foo); //outputs: {name: 'foo'}
instance.callHook('pre:whatever', function(){
  console.log(foo);// outputs: {name: 'changed by middleware'}
});
```

> Wordpress lets you do this, and it has contributed in a major way to the power of using it as a platform

Yeah, but this is not Wordpress. I think you're a bit fixated on what you already know. It takes some getting used to, but in node-land things happen a bit differently. Sometimes for the better, sometimes for the worse. (The latter being the things I'm trying to change too: for instance make DI-libs common enough) But anyway, there's plenty of stuff that's really cool and powerful, yet requires a bit of a different way of thinking. I come from a PHP and AS3 background and it took me quite some time to get used to how all of this works.

**Comment by mattapperson** [comment]
@creynders fair enough, my contrived example expected something different but you are clearly correct that this does work. I personally find the code a little hard to follow (why is what I set there changed???) but it does work so this is I suppose a matter of preference.

As for my wordpress analogy... Please don't assume i am just a wordpress refugee. I have been developing using node for many years, my argument does not come from what I am used to, but rather a desire to enable keystone to be flexible and powerful enough to attract as many devs as wordpress did/does. 
I 100% understand that this is perhaps not best practice, but at the same time, it gives a level of power and flexibility that if lacking results in being forced to look elsewhere to other platforms.
<rant/>
At the end of the day though, this is your baby, your project. I am just throwing my $0.02 in the ring

**Comment by creynders** [comment]
> why is what I set there changed???

Since data isn't immutable in JS and the object is passed by reference, not by value. I.e. the `foo` inside the middleware references the exact same object as the `foo` in the callback.

> it gives a level of power and flexibility that if lacking results in being forced to look elsewhere to other platforms

Yeah, sure. But there's many ways to skin a cat. A real plugin architecture, like Wordpress's, is something which needs to be mapped out still in keystone. Once we start on that I'll definitely try and remember to give you a shout out. 

> this is your baby, your project

Not sure whether you're talking about keystone, but I'm just one of the contributors there, though I do think it's a very cute baby ;) 

And your $0.02 are more than welcome. I hope you don't think I'm not open to paradigm or viewpoint shifts, I am. But I always try to find out if we're not trying to solve a problem that doesn't really exist.

**Comment by mattapperson** [comment]
@creynders While you are correct, and logically I know this... its sort of a JS "WUT!" moment IMHO. Plus the fact that Keystone is adopting React and Redux points to a community that will increasingly expect immutable data flows. So for the server side to embrace mutable data and the Client to shun it... I think this will create a major pain point for divs trying to learn and use the platform. The UI is going immutable, the back end should too IMHO as much as possible.

To your point on plugins being mapped out... that's kind of what I am trying to get the ball rolling on. For right now this is basically keystones "internal" plugin arch more or less... so I thought here would be a good place to start.
My mentality when designing APIs such as this have always been. Provide a simple, clean API that has a small service area. Give divs all the power to jack things up or build amazing things, just warn them when doing something stupid (even if only in "debug" mode)

I was speaking to grappling-hook as your baby... Cut I mean like... you have seriously just written and maintained it almost single handedly 😊

**Comment by creynders** [comment]
> So for the server side to embrace mutable data and the Client to shun it... [...] The UI is going immutable, the back end should too IMHO as much as possible.

Gotcha. And I even somewhat agree. I don't have enough experience with React to have a strong opinion on it, but it looks promising. Once we go down that road (i.e. immutable data on the backend) we'll have to come up with different solutions altogether and forget about middleware I think. However there's still too many unknown unknowns ATM to have a clear idea of where we'll head IMO.

**Comment by mattapperson** [comment]
I don't think it's bad to have middleware, redux itself uses / has a middleware API. You would just need to move past the express frameworks style of a middleware API. And that's kinda what I am arguing for here. Like you said, there is more then one way to skin a cat 😊

**Comment by creynders** [comment]
Yeah, sorry that's what I meant, the express-style middleware. The thing is: as it is now, it's very versatile. For instance we _do_ use it to register middleware to express routes as well. Which makes sense, since the express app is a fundamental part of keystone as well. I.e. it's good to keep close to the other stuff you're already using anyway. But there's plans to make express optional, i.e. at a certain point in time it probably _won't_ make sense to stick to it any more. Whence I'll probably come back to this discussion to think on a different approach for middleware. But whether it'll be closer to React, or ...?

**Comment by mattapperson** [comment]
Might I argue then that if we know that this will happen at some point and with a rapidly growing community... it is better to introduce new APIs ASAP and deprecate the old ones. Otherwise all your doing is allowing more technical debt to build up for users of the platform.

**Comment by creynders** [comment]
Hehe, yeah. The amount of things to do ASAP is truly staggering. Seriously, we've spent the last 9 months on converting the admin to an SPA. That's it. There's tons and tons of stuff that need to happen before this even blips at the radar. 
And personally, keystone's just one of 40+ OSS projects I've committed myself to. I'll do stuff when it's needed, not to accomodate what-ifs 😁

**Comment by mattapperson** [comment]
@creynders im not trying to push you to do the work yourself or to rush. This is all OSS after all, not like it's your day job. I would just like to see if we could get to an agreed on spec (or even just direction) that the community could contribute to. If it's too much for right now though, then that is what it is.

**Comment by creynders** [comment]
> I would just like to see if we could get to an agreed on spec

We can certainly try. If you mean the whole plugin architecture, I'll list some related issues in keystonejs/keystone#1860
Or specifically for grappling-hook, we can discuss it further here, or in a new issue.

**Comment by creynders** [comment]
@mattapperson closing this, but if you do want to discuss a spec, let's open a new issue and link back.
